### ISSUE TITLE
Overlay - Element events (click) pass through to Map

### ISSUE THREAD

**Issue by ralphstodomingo** [openlayers/openlayers#6948]
I have a popup in React made as an Overlay. The problem is, events meant for this React component (clicking a button) pass through to the map, which means sometimes features get (de)selected because the buttons happen to be over them.

Setting the overlay with `stopEvent: true` causes the React component to not receive any click events at all, while `false` causes the map to be also clicked as well.

Is this intended behavior? Does 'stopEvent' prevent the overlay element from having any events?

**Comment by bartvde** [comment]
yeah I've dealt with this as well, there are multiple threads on it such as:

https://stackoverflow.com/questions/35454014/clicks-on-reactjs-components-not-firing-on-an-openlayers-overlay

I've also used the approach of setting interactions to not active when the popup is open such as here: 

https://github.com/boundlessgeo/sdk/blob/3ee2e70c6c35d3390be4804058ee9c0521da7806/src/components/EditPopup.js#L223:L228

See also: https://github.com/openlayers/openlayers/issues/4400

**Comment by tschaub** [comment]
I think we'd agree that setting `stopEvent: true` should only cause the map to ignore events that target the overlay.  We should be able to handle this in the library, so I think it is a bug (instead of intended behavior) that other listeners can't respond to events on the overlay.

**Comment by ralphstodomingo** [comment]
For my particular use-case my workaround involved setting `stopEvent: true` and doing something like this inside my React component (which I took from @bartvde's answer in that linked SO question):

```
if (this.closeButtonRef.button.button.onclick === null) {
   this.closeButtonRef.button.button.onclick = this.setFeaturePopupVisibleAsFalse;
}
```
 
This is because I was using `material-ui` which comes with its own buttons. A normal button will likely involve `this.buttonElement.onclick = this.testFunc` or something similar.
I also had to bind this in `componentWillReceiveProps` instead of the constructor or `componentDidMount` since the element in question relied on a prop triggering before appearing.
This looks hacky at best, and I had to do this for every action button on that popup. Still, hoping this helps anyone who stumbles with it, while there isn't a fix yet.

**Comment by bartvde** [comment]
Yep would be great if we could handle this in the library instead, I agree those workaround are wonky ;-)

**Comment by marcjansen** [comment]
We have deployed such wonky workarounds as well, I would also favor a library solution.

**Comment by bassarisse** [comment]
Any news on this? Looking for a better solution, at least without having to patch each component =/

**Comment by hwbllmnn** [comment]
@tschaub @ahocevar I'd be willing to work on a solution for this. Can you give me a few pointers on how to approach this? Thanks

**Comment by marcel-dancak** [comment]
Sill no official solution, so I will add another workaround for React. You can enable it simple on overlay's root element, and onClick events should be working almost normally in all nested elements. It works by converting _mouseup_ events to _click_ events with overridden _stopPropagation_ method, so ol cannot stop it.

```javascript
const convertToClick = (e) => {
  const evt = new MouseEvent('click', { bubbles: true })
  evt.stopPropagation = () => {}
  e.target.dispatchEvent(evt)
}
```
```jsx
<div onMouseUp={convertToClick}>
  <button onClick={handleClick}>x</button>
  ...
</div>
```

(For simple use cases, it is easier to directly use onMouseUp instead of onClick)

**Comment by benupham** [comment]
I have been having trouble with event propagation and OpenLayers as well. I'm not sure if this is relevant to React, or the best solution, but I wanted to point out (no pun intended) that the click event that the ol map listens for it "pointerdown" not "click" -- at least in some cases. I was stopping propagation on "click" on an overlay, which wasn't working, but stopping propagation on "pointerdown" does. Similar for "mousemove" -- ol listens for "pointermove".

**Comment by dnyanp** [comment]
Any update on this?

**Comment by walkermatt** [comment]
I've recently encountered this when testing ol-geocoder and ol-layerswitcher components which don't use React but do remove or recreate the elements that the events originate from.

I think the issue is that once the elements have been removed [the check to detect whether an event originated from an element that is a child of `ol-overlaycontainer-stopevent`](https://github.com/openlayers/openlayers/blob/master/src/ol/PluggableMap.js#L936-L942) no longer holds as the elements are no longer in the DOM.

I think a fix might be to check if the `target` is in the page in `ol/PluggableMap~PluggableMap#handleMapBrowserEvent` like so:

```diff
--- a/src/ol/PluggableMap.js
+++ b/src/ol/PluggableMap.js
@@ -950,8 +950,23 @@ class PluggableMap extends BaseObject {
       return;
     }
     let target = mapBrowserEvent.originalEvent.target;
+    if (!document.body.contains(target)) {
+        return;
+    }
     while (target instanceof HTMLElement) {
```
Happy to submit a PR if acceptable.

**Comment by walkermatt** [comment]
[Demo app to reproduce the issue](https://codesandbox.io/s/custom-controls-stop-event-synw5). If you open the console you should see `singleclick` events logged when clicking on the N/S button. This is due to the `RotateNorthControl` control rebuilding the UI once the user clicks the button. The example is a little contrived but it demonstrates the issue. I've found that the above patch :point_up: resolves the issue when applied to a local copy of the repo.

**Comment by walkermatt** [comment]
Hi @ahocevar, is this issue a candidate for being reopened or would I be better creating a new issue?

**Comment by ahocevar** [comment]
@walkermatt I'm reopening this. Your fix would need some comments, but if it also solves the reported issues with React I'd not be opposed.

**Comment by walkermatt** [comment]
@ahocevar I've pulled together an [example that uses React to provide the UI for a Control](https://codesandbox.io/s/hjkg6) which again reproduces the issue. There is a singleclick handler on the map which is triggered when the show/ hide buttons are clicked (see the console log). My local copy which includes the change to PluggableMap fixes the problem.

**Comment by ahocevar** [comment]
@walkermatt I think a React Control example would be a nice addition. Thanks in advance!

**Comment by jumpinjackie** [comment]
The fix introduced with #10502 and shipped with OpenLayers 6.2 breaks all mouse events in my map viewer. I am in a bit of an inverted situation. I am not trying to use React components as controls/overlays for an OL map. Instead I have a React component that wraps the OL map and in this situation I lose all mouse events with OL 6.2 so I cannot pan, mouse wheel zoom, drag, etc, etc.

The react component is quite complex so I don't know what specifics I should provide to give better context for diagnosing/debugging the issue. But as a starting point to at least demonstrate the problem, I've got a storybook instance on my GH pages repo that demonstrates the problem:

https://jumpinjackie.github.io/mapguide-react-layout/olbug/storybook-static/index.html

Choose any story that showcases the map. You should notice that mouse based actions (pan, mouse wheel zoom, etc) do not work at all.

Compare this with the current master, which is also updated to OL 6.2, but I've [monkey-patched out the bad PluggableMap method](https://github.com/jumpinjackie/mapguide-react-layout/blob/master/src/api/bootstrap.ts#L10).

https://jumpinjackie.github.io/mapguide-react-layout/master/storybook-static/index.html

All the mouse based actions work here.

**Comment by ahocevar** [comment]
@jumpinjackie, please make sure you include `css/ol.css` from v6.2.0 with your component. Looks like the `pointer-events: none` style is missing from the overlay container divs.

**Comment by jumpinjackie** [comment]
Thanks @ahocevar that does indeed fix the issue. I had a stale local pre-6.0 OL stylesheet.

**Comment by ZakarFin** [comment]
> The fix introduced with #10502 and shipped with OpenLayers 6.2 breaks all mouse events in my map viewer. 

This was a life-saver comment. I've been debugging our React/OpenLayers app for some time now trying to figure out which library upgrade broke the map. Turns out we haven't included ol.css at all previously using ol 6.1.1 and older. Looks like adding an import for the ol/ol.css fixed the issue.

**Comment by Voronar** [comment]
Is this new React release helps with the issue?
https://reactjs.org/blog/2020/08/10/react-v17-rc.html#changes-to-event-delegation

**Comment by namedgraph** [comment]
I'm struggling to understand what `ol.Overlay` does with the supplied `options.element`. It seems that the user-supplied element becomes a child of `.ol-overlay-container ol-selectable`, but I cannot figure out where/how that happens.

It doesn't help that `this.element` seems to be the container element while `this.getElement()` is the user-supplied element?

I'm handling `onclick` within an overlay element and it seems to propagate to an out-of-DOM orphan clone of the `options.element` I had supplied to `ol.Overlay`. Any idea how that can happen?

**Comment by namedgraph** [comment]
Should I be appending my overlay element directly to the `map.getOverlayContainerStopEvent()` element?
